distutils/unixccompiler.py

```diff
@@ -196,7 +196,12 @@ def link(self, target_desc, objects,
                     else:
                         offset = 0
 
-                    linker[i+offset] = self.compiler_cxx[i]
+                    if len(linker) >= len(self.linker_exe) and \
+                            linker[:len(self.linker_exe)] == self.linker_exe:
+                        linker = linker[:(i + offset)] + self.compiler_cxx + \
+                            linker[len(self.linker_exe):]
```

Comment:
I'm trying to understand what this logic is attempting to accomplish. It touches so many things (linker, self.linker_exe, self.compiler_cxx, i, offset) and it doesn't seem to be very congruent with the `else` branch, which simply substitutes a single element of `linker`. This change is adding quite a bit of complexity to an already complex function (increases the mccabe cyclometric complexity from 15 to 16, where a target of 10 or less is preferred).

Help me understand - when is `len(linker) >= len(self.linker_exe)`? By my reading of the code, that should only happen when `linker = self.linker_so[:]` (from line 179 above). Is that right?

Would it be possible to add some comments that explain what's going on here and what are all the different factors affecting its behavior?